src/dbms/inmemory/replication_handlers.cpp

```diff
@@ -472,16 +478,15 @@ uint64_t InMemoryReplicationHandlers::ReadAndApplyDelta(storage::InMemoryStorage
                       delta.edge_create_delete.from_vertex.AsUint(), delta.edge_create_delete.to_vertex.AsUint());
         auto *transaction = get_transaction(timestamp);
         auto from_vertex = transaction->FindVertex(delta.edge_create_delete.from_vertex, View::NEW);
-        if (!from_vertex) throw utils::BasicException("Invalid transaction!");
+        if (!from_vertex) throw utils::BasicException("Invalid transaction! Please raise an issue, line: {}", __LINE__);
         auto to_vertex = transaction->FindVertex(delta.edge_create_delete.to_vertex, View::NEW);
-        if (!to_vertex) throw utils::BasicException("Invalid transaction!");
-        auto edges = from_vertex->OutEdges(View::NEW, {transaction->NameToEdgeType(delta.edge_create_delete.edge_type)},
-                                           &*to_vertex);
-        if (edges.HasError()) throw utils::BasicException("Invalid transaction!");
-        if (edges->edges.size() != 1) throw utils::BasicException("Invalid transaction!");
-        auto &edge = (*edges).edges[0];
-        auto ret = transaction->DeleteEdge(&edge);
-        if (ret.HasError()) throw utils::BasicException("Invalid transaction!");
+        if (!to_vertex) throw utils::BasicException("Invalid transaction! Please raise an issue, line: {}", __LINE__);
+        auto edgeType = transaction->NameToEdgeType(delta.edge_create_delete.edge_type);
+        auto edge =
+            transaction->FindEdge(delta.edge_create_delete.gid, View::NEW, edgeType, &*from_vertex, &*to_vertex);
```

Comment:
I would assume this FindEdge is quite costly if we're deleting from a supernode?
Does it make sense to probe the source out edges as well as destination in edges in order to choose the one with less edges?

Reply:
WalDeltaData::Type::EDGE_DELETE was previously implemented in terms of OutEdges. The new FindEdge is implemented in terms of OutEdges. Don't expect a performance degradation as previously replication didn't allow multiple edges of same type.

Also filter of a `std::vector<std::tuple<EdgeTypeId, Vertex *, EdgeRef>>` I doubt is a major expense even for a supernode.

You are right, performance could be improved. TODO: `FindEdge` internals, check both to and from vertex.

Reply:
Doing a quick check to see who has fever edges and then using that node for the search.